coldfront/api/project/views.py

```diff
@@ -15,3 +36,69 @@ class ProjectViewSet(mixins.ListModelMixin, mixins.RetrieveModelMixin,
 
     def get_queryset(self):
         return Project.objects.all()
+
+
+class ProjectUserRemovalRequestViewSet(mixins.ListModelMixin,
+                                       mixins.RetrieveModelMixin,
+                                       mixins.UpdateModelMixin,
+                                       viewsets.GenericViewSet):
+    """A ViewSet for the IdentityLinkingRequest model."""
+
+    filterset_class = ProjectUserRemovalRequestFilter
+    http_method_names = ['get', 'patch']
+    permission_classes = [IsSuperuserOrStaff]
+    serializer_class = ProjectUserRemovalRequestSerializer
+
+    def get_queryset(self):
+        return ProjectUserRemovalRequest.objects.order_by('id')
+
+    @swagger_auto_schema(
+        manual_parameters=[authorization_parameter],
+        operation_description=(
+                'Updates one or more fields of the ProjectUserRemovalRequest '
+                'identified by the given ID.'),
+        auto_schema=None)
+    @transaction.atomic
+    def partial_update(self, request, *args, **kwargs):
+        """The method for PATCH (partial update) requests."""
+        logger = logging.getLogger(__name__)
+
+        partial = kwargs.pop('partial', False)
+        try:
+            instance = self.get_object()
+            serializer = self.get_serializer(
+                instance, data=request.data, partial=partial)
+
+        except Http404:
+            serializer = self.get_serializer(
+                data=request.data, partial=partial)
+
+        try:
+            serializer.is_valid(raise_exception=False)
```

Comment:
That was left over from testing

Reply:
Got it. We should move this line outside of the `try`-`except`, as its errors should be forwarded back to the user automatically, without having to explicitly return them.

Since we've established that the serializer is valid (i.e., user input is valid), any errors raised (e.g., by the runner) are server errors, so it'd be appropriate to return a 500 status code rather than a 400 (admittedly, the API is not 100% consistent about this).

We should also handle errors raised within the runner (any exception raised in `complete_request` is caught within it, so it wouldn't get detected by the caller).

==========

A broader TODO for me that I should have done from the start is to wrap all these multi-step operations in transactions all over the codebase, so that they happen in an all-or-nothing fashion, preventing intermediate states.

For example, suppose the code successfully marks the request as "Complete", but an exception occurs when actually removing the user from the project. It won't be processed again because it's "Complete", but we haven't actually done everything we needed to do.

Because we've decorated this API call with `@transaction.atomic`, it would roll back changes if an uncaught exception occurs. (In this case, it won't, because the exception is caught.)

Reply:
I'm currently working on making the processing atomic.